astropy/coordinates/representation.py

```diff
@@ -255,6 +255,19 @@ def _apply(self, method, *args, **kwargs):
                     apply_method(getattr(self, component)))
         return new
 
+    def _setitem(self, item, value):
```

Comment:
My sense would be to move this method to `BaseFrame`, as I don't think `Representation` has to know about this.

Or implement `__setitem__` for representations and (somehow) take care that `coord.data` contains or returns a read-only version.

Reply:
This cannot easily be moved out to `BaseFrame` because of the differentials.  Notice that there are two versions of `_setitem` and the differentials one calls the `super()` version.  So these really are part of the `Representation` class structure (in that they deal with different behavior of subclasses).

On the other idea of making the representation `data` be read-only, that pretty much breaks everything, including the existing (as of 4.0) API for updating coordinates in place.

Reply:
Hmm, fair point about the subclassing.

But I think I wasn't very clear about my alternative suggestion: it was to just define `__setitem__` with what you have here, since representations themselves have no reason to be immutable (they're just complicated quantities as far as I'm concerned, and your current `_setitem` ensures we do not set with a wrong representation class).

The read-only aspect I mentioned only because, as far as I can tell, the only reason representations are currently immutable is to ensure people do not mess up by setting data. But as you note, we in fact show work-arounds for that, so if we are making `SkyCoord` mutable anyway, perhaps we do not have to worry too much about `sky_coord.data` becoming mutable itself.

Reply:
@eteq, @taldcroft - thinking more about this, I still think we should just give representations a `__setitem__` - they really have no reason to be immutable. On the caching question, I think it is important that we make sure our code works well, and that we have appropriate warnings to avoid user errors; I don't think we should make convoluted code just to ensure users cannot make mistakes. So, I would suggest to use `__setitem__` for the representations as well as for `SkyCoordinate` and `BaseFrame`, and be sure to add a warning to the docstring of `data`.

p.s. Just as an example of how hopeless it is to prevent users messing up if they go for attributes, consider a user who, quite logically in my opinion (arguably more logical at least than setting `data`), thinks that one can change coordinates from their named attributes:
```
sc = SkyCoord(np.arange(2.)*u.deg, np.arange(2.)*u.deg)                                        
sc                                                                                             
# <SkyCoord (ICRS): (ra, dec) in deg
#     [(0., 0.), (1., 1.)]>
sc.ra                                                                                          
# <Longitude [0., 1.] deg>
# Aha, I can just change RA, right?
sc.ra[1] = -1*u.deg                                                                            
# Let me just check to be sure
sc.ra                                                                                          
# <Longitude [  0., 359.] deg>
# Clearly, that worked.
# But...
sc                                                                                             
# <SkyCoord (ICRS): (ra, dec) in deg
#     [(0., 0.), (1., 1.)]>
```

Reply:
@mhvk - indeed the current infrastructure of frame / representation and the dependence on caching (for performance and user-experience reasons) leaves coordinates vulnerable to problems like your example.  That is an important separate discussion, but completely out of scope for this PR, which is just adding one safe way to set items.

What is the intended public-facing use-case for having `__setitem__` on representations?  Why do we need to expose that, and in the process expose an additional way of accidentally corrupting a coordinate object?

I don't think that calling a method `_setitem` vs. `__setitem__` changes the code complexity much.  I could move all of that code into `BaseFrame.__setitem__` but that would actually make the code more complex by making that method a lot bigger, with handling that is specific to the representation not the frame.

Reply:
I wonder if the conversation would be helped by describing the `.data` representation as "protected" instead of mutable or immutable.  The idea of "protected" is that users have to take unusual steps to update `.data` in place, in this case explicitly clearing a cache that is not directly part of the representation object itself.  The point of not providing `__setitem__` is to help the users to get to the documentation pages which describe the dragons.

Reply:
I like "protected"!

My main reason for thinking representations should have `__setitem__` is that that is the standard way to set items in python, and if we allow setting in `SkyCoord` and the frames using `__setitem__` it just seems illogical not to do the same for representations. Why should the lower level be done in a different way? What makes them special?

I think this in particular since the representations are not only useful for coordinates, but also outside, as a general way to represent 3-d vectors (as we document; http://docs.astropy.org/en/latest/units/quantity.html#representing-vectors-with-units and http://docs.astropy.org/en/latest/coordinates/representations.html#vector-arithmetic). They much behave like quantities, so might as well be similarly settable.

Reply:
p.s. Of course, none of this would matter if we didn't have to have caching... https://www.martinfowler.com/bliki/TwoHardThings.html

Reply:
>  Why should the lower level be done in a different way? What makes them special?

It is exactly the caching and the fact that `.data` does not have any reference to a parent frame, so it cannot invalidate the cache in `__setitem__`.  Caching is basically needed because of APE-5 and https://github.com/astropy/astropy-api/pull/12, e.g.
```
assert icrs.spherical.lat == 5*u.deg
assert icrs.cartesian.z.value > 0
icrs.representation_type = 'cartesian'  # It works
```
In other words the API allows seamlessly changing representations, so caching is needed to make that performant. I will accept some blame here, this seemed like a fine idea and the community accepted it.  If we had a time machine it would be worth trying to avoid anything in the API that requires caching.  But that ship has sailed.  I don't see a way out without make a new class or a serious API break.

Having said all this, can we open a new issue to "consider adding `__setitem__` to representations", and stipulate that this is out of scope for *this PR*, which is explicitly about making SkyCoord and Frame a little better and to support table operations for SkyCoord. In other words, perfect is the enemy of good.

Reply:
Just to be clear, I am not claiming caching is a bad idea, just remarking that it is known in computer science that to do it right is very hard (Indeed, that has already been amply demonstrated for `Time`, where it similarly makes sense to cache, but where of course setting the equivalent of `.data`, `jd1[...]` and `jd2[...]` screws up the cache as well).

But before just giving in on making a separate issue (which just means I have to make the changes and we will need to have the whole discussion all over again): what is really wrong with just changing `_set_item` to `__setitem__`? Right now one can screw up nearly everything else in a `SkyCoord` by setting an item in most attributes (except `TimeAttribute`, which is made read-only; quantities, locations, and now coordinate origins are settable), so why make `.data` different. In fact, arguably doing so is confusing since not allowing setting it suggests it is safe to set all the other attributes.

And of course, by changing to `__setitem__`, table would work for representations as well! 

Note that there is *no* additional risk of previous code breaking; we are only talking about new users possibly messing up.

p.s. I'm quite willing to make a PR to yours that does it, adjusts the `.data` docstring, and adds some tests for representations.

Reply:
What is really the issue is that making the change is *controversial*, so I do not want my PR being blocked and delayed by a controversial discussion that has nothing to do with the goal of this PR.  That change is out of scope and so it really belongs in a separate PR.  We should keep each PR as small as possible, no?  

This one is done and ready to merge now (apart from one missing backtick and final review of doc updates from @eteq), and there is no reason it needs a new PR to make it complete or even better.  Once this is merged (which could happen as soon as @eteq looks over changes), then you can make a nice clean PR against master.

Reply:
@taldcrof - OK, let's indeed go that route. I've gotten quite convinced this all is a good idea, and, as you say, we can do the next step in a separate PR.